// clean-css-brunch/lib/index.js
// Generated by CoffeeScript 1.6.2
var CleanCSSMinifier, cleanCSS, sysPath,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

cleanCSS = require('clean-css');

sysPath = require('path');

module.exports = CleanCSSMinifier = (function() {
  CleanCSSMinifier.prototype.brunchPlugin = true;

  CleanCSSMinifier.prototype.type = 'stylesheet';

  function CleanCSSMinifier(config) {
    var _ref, _ref1, _ref2;

    this.config = config;
    this.optimize = __bind(this.optimize, this);
    this.options = (_ref = (_ref1 = this.config) != null ? (_ref2 = _ref1.plugins) != null ? _ref2.cleancss : void 0 : void 0) != null ? _ref : {};
  }

  CleanCSSMinifier.prototype.optimize = function(data, path, callback) {
    var err, error, optimized;

    try {
      optimized = cleanCSS.process(data, this.options);
    } catch (_error) {
      err = _error;
      error = "CSS minify failed on " + path + ": " + err;
    }
    return process.nextTick(function() {
      return callback(error, optimized || data);
    });
  };

  return CleanCSSMinifier;

})();
